uni/ulsp/formatting.icn

```diff
@@ -0,0 +1,354 @@
+            if ch == " " then {
+               if spaces == 0 then
+                  position := &pos - 2
+               spaces := 1
+            } else spaces := 0
+         }
+      }
+      if spaces = 1 then
+         return createResult(lineNum - 1, position, lineNum - 1, *_line, "")
+   end
+
+   #<p>
+   #   Checks the indentation and returns a result fixing any errors based on the
+   #   level of the line and the tab size set in VsCode.
+   #   <i>Used internally.</i>
+   #   <[param _line The actual content of the line.]>
+   #   <[param lineNum The line number to check.]>
+   #   <[param tabAmount The level of the line multiplied by the tab size.]>
+   #   <[returns the line with the proper amount of indentation.]>
+   #</p>
+   method fixIndentation(_line, lineNum, tabAmount)
+      local correctSpaces, actualSpaces, ch
+
+      correctSpaces := repl(" ", tabAmount)
+      actualSpaces := 0
+      _line ? {
+         while ch := move(1) do {
+            if ch == "#" then
+               return
+            if ch ~== " " then
+               break
+            actualSpaces +:= 1
+         }
+         if actualSpaces ~= tabAmount then
+            return createResult(lineNum - 1, 0, lineNum - 1, &pos - 2, correctSpaces)
+      }
+   end
+
+   #<p>
+   #   Looks through each operator in the line and adds spaces before/after when appropriate.
+   #   <i>Used internally.</i>
+   #   <[param _line The actual content of the line.]>
+   #   <[param lineNum The line number to check.]>
+   #   <[returns the line with corrected spacing between operators.]>
+   #</p>
+   method insertSpaces(_line, lineNum)
+      local res, charOps, position
+
+      res := []
+      charOps := ':=+-/*^%~<>?|-@'
+
+      _line ? {
+         while tab(upto(charOps)) do {
+            if withinQuotes(_line, &pos) = 0 then {
+               case _line[&pos] of {
+                  ":" : {
+                     if _line[&pos - 1] ~== " " then {
+                        position := &pos
+                        if *tab(many(charOps)) > 1 then {
+                           put(res, createResult(lineNum - 1, position - 1, lineNum - 1, position - 1, " "))
+                           if _line[&pos] ~== " " then {
+                              if _line[&pos - 1] == "-" && _line[&pos - 2] == "=" then
+                                 put(res, createResult(lineNum - 1, &pos - 2, lineNum - 1, &pos - 2, " "))
+                              else
+                                 put(res, createResult(lineNum - 1, &pos - 1, lineNum - 1, &pos - 1, " "))
+                           }
+                        } else
+                           tab(many(charOps))
+                     } else {
+                        tab(many(charOps))
+                        if _line[&pos] ~== " " then {
+                           if _line[&pos - 1] == "-" && _line[&pos - 2] == "=" then
+                              put(res, createResult(lineNum - 1, &pos - 2, lineNum - 1, &pos - 2, " "))
```

Comment:
seems like every use of `lineNum`  is  `lineNum - 1` , why not pass `lineNum - 1` to the function in the first place if that what expects? or at least do `lineNum  -:= - 1` at the top of the function.